Extend parse function to allow multiple values per key

gassmoeller opened the pull request:
This PR extends Utilities::parse_map_to_double_array to allow multiple entries per key using the syntax: `key1: value1|value2|value3, key2: value1, key3: value1|value2`. We would like to use this syntax to make progress with defining phase transitions for multiple compositional fields (like #2656). The unit tests should cover all cases, so I can make this as a separate PR without actual implementations using it (it will be used by several rheologies, equations of state and material models). 

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.

gassmoeller commented:
@jdannberg, @naliboff, @mibillen can one of you take a look if this is what we talked about?

naliboff commented on source/utilities.cc:
```diff
@@ -230,7 +254,7 @@ namespace aspect
                                    + property_name
                                    + "> was not found. Specify a comma separated "
-                                   "list of `<double>' or `<id> : <double>'."));
+                                   "list of `<double>' or `<id> : <double>|<double|...'."));
```
```suggestion
                                   "list of `<double>' or `<id> : <double>|<double>|...'.")); 
```
Maybe a missing `>`?

gassmoeller commented:
Thanks for the quick review. I have fixed the comment and added one more functionality we will need. That is to return the actual substructure of the returned vector. As discussed we need this to ensure that all entries actually have the same number of phases for each individual composition (e.g. to distinguish `a:1|2,b:3` from `a:1,b:2|3`).

jdannberg commented on include/aspect/utilities.h:
```diff
@@ -79,7 +79,11 @@ namespace aspect
+     * @p list_of_keys.
      *
      * This function also considers a number of special cases:
      * - If the input string consists of only a comma separated
```
Do you want to add  comment here that mentions that you can not do this if `allow_subentries == true` because then the notation does not provide enough information?

jdannberg commented on include/aspect/utilities.h:
```diff
@@ -110,6 +114,14 @@ namespace aspect
+     * @param [in] allow_subentries If true allow having multiple subentries
+     *   for each key. If false only allow a single entry per key. In either
+     *   case each key is only allowed to appear once.
+     * @param [out] subentry_structure If handed over this
```
To me, this name does not seem as descriptive as it could be. What about something like `values_per_key` or `n_values_per_key`? That would also be consistent with the naming of the other input parameters.

jdannberg commented on unit_tests/parse_map_to_double_array.cc:
```diff
@@ -82,6 +82,36 @@ TEST_CASE("Utilities::parse_map_to_double_array")
+    INFO("check 10: ");
+    auto subentry_structure = std::make_shared<std::vector<unsigned int>>();
+
+    compare_vectors_approx(aspect::Utilities::parse_map_to_double_array ("C1:100, C2:200|, C3:300, C4:400, C5:500",
```
Oh, we actually allow to use the | but then not have an entry? Is that intended?

gassmoeller replied:
That is allowed by deal.II (it depends on the implementation of the `split_string_list` function, which explicitly allows this case). I think this is fine, likely nobody wants to use it.

jdannberg commented on unit_tests/parse_map_to_double_array.cc:
```diff
@@ -135,6 +165,32 @@ TEST_CASE("Utilities::parse_map_to_double_array FAIL ON PURPOSE")
+  // Subentries not allowed
+  INFO("check fail 7: ");
+  REQUIRE_THROWS_WITH(
+    aspect::Utilities::parse_map_to_double_array ("C1:100, C2:|200, C3:300, C4:400, C5:500",
```
Wouldn't this test be more helpful and easier to read if the output format was correct (and only the subentries are not allowed)? This way, how do you know which of the two it is?

jdannberg commented on source/utilities.cc:
```diff
@@ -206,22 +231,40 @@ namespace aspect
+            for (const auto &sub_entry: entry)
+              return_values.push_back(sub_entry);
         }
       else if (Patterns::List(Patterns::Double(),1,n_fields).match(input_string))
```
We should make sure that `allow_subentries` is false in this case.

gassmoeller replied:
I would not want to do that. Consider the case that somebody wants to allow multiple values per key, but does not know the input structure the user puts into the .prm file. You would still want to allow a simple list of only a single value per key. The parameter is `allow_multiple_values_per_key`, not `enforce_multiple_values_per_key`. But you are right, I should document that this format expects exactly one value per key (and we already enforce that, because `possibly_extend_from_1_to_N` expects either 1 entry, or `n_fields` entries).

jdannberg commented on source/utilities.cc:
```diff
@@ -89,21 +89,41 @@ namespace aspect
+                                              "|")
+                               :
+                               Patterns::Double()
+                               ;
```
Move that to the end of the line before?

jdannberg commented on source/utilities.cc:
```diff
@@ -146,7 +166,12 @@ namespace aspect
 
```
There is a space missing in line 162 above.

jdannberg commented on source/utilities.cc:
```diff
@@ -146,7 +166,12 @@ namespace aspect
+
+                      if (subentry_structure)
+                        (*subentry_structure)[field_index] = 1;
+                    }
```
So I guess that mean we don't allow entries in the format `all:100|200`. Is that something we want, or, alternatively, should we state somewhere you can't do that?

jdannberg commented on include/aspect/utilities.h:
```diff
@@ -110,6 +114,14 @@ namespace aspect
+     *   entries is the length of the returned vector of values.
      *
      * @return A vector of values that are parsed from the map, provided
      *   in the order in which the keys appear in the @p list_of_keys argument.
```
You should add some documentation here that explains the new structure with subentries. That could also be a good place to explain why the `subentry_structure` is useful (it means whoever calls the function can figure out which value belongs to which key).

gassmoeller commented:
I think I addressed all comments, and also added the functionality to hand over an existing vector of `n_values_per_key` that can be used as the expected template for the input parameter. This saves us from implementing Asserts for every parameter to check that its structure is the same as the last one that was parsed. Every change is in the last commit.

jdannberg commented on source/utilities.cc:
```diff
@@ -94,27 +94,37 @@ namespace aspect
                                const std::vector<std::string> &input_field_names,
```
For some reason, the names of the first two function arguments are different in the declaration and in the implementation here.

jdannberg commented on source/utilities.cc:
```diff
@@ -260,10 +287,18 @@ namespace aspect
-                (*subentry_structure)[i] = 1;
+                AssertThrow((*n_values_per_key)[i] == 1,
+                            ExcMessage("The input parameter " + property_name + " does not have "
+                                       + "the expected number of values for field index " + std::to_string(i) + "."));
```
I find this error message a bit confusing. If this error is thrown, it means someone specified a list of doubles in the input file with one value per key, but some other parameter was specified with multiple values per key, provided in the key:value format (or there is some other reason why the function expects more than one value per key). 

So to help the user a bit more, I would tell the user how many entries were expected, and tell them something like: "To specify more than one value per field for this input parameter, you need to use the format <key>:value1|value2|..."

gassmoeller commented:
I addressed the comments. Thanks for the review!
